python/lsst/meas/extensions/scarlet/io.py

```diff
@@ -472,3 +482,153 @@ def oldScarletToData(blend: Blend, psfCenter: tuple[int, int], xy0: Point2I):
+        return result
+
+    # Extracting a single file from the ZIP
+    def _extract_blend(self, uri: ResourcePath, blend_filename: str) -> dict[str, Any]:
+        """Extract a single file from a ZIP archive.
+
+        Parameters
+        ----------
+        uri : `ResourcePath`
+            URI of the ZIP file
+        blend_filename : `str`
+            Name of the blend file to extract
+
+        Returns
+        -------
+        blend :
+            Dictionary mapping filenames to the desired file type.
+        """
+        results = {}
+        filenames = [
+            blend_filename,
+            'psf.json',
+            'psf_shape.json',
+        ]
+        with uri.open("rb") as f:
+            with zipfile.ZipFile(f, 'r') as zipf:
+                for filename in filenames:
+                    with zipf.open(filename) as f:
+                        results[filename] = from_json(f.read())
+        return results
+
+    def _build_model(self, zip_dict: dict[str, Any]) -> scl.io.ScarletModelData:
+        """Build a ScarletModelData instance from a dictionary of files.
+
+        Parameters
+        ----------
+        zip_dict : `dict` of `str` to `object`
+            Dictionary mapping filenames to the desired file type.
+
+        Returns
+        -------
+        model :
+            ScarletModelData instance.
+        """
+        model_psf = zip_dict['psf.json']
+        psf_shape = zip_dict['psf_shape.json']
+        blends = {}
+        for key, value in zip_dict.items():
+            if key.startswith('blend_'):
+                blend_id = int(key.split('.')[0].split('_')[1])
+                blends[blend_id] = value
+        return scl.io.ScarletModelData.parse_obj({
+            'psf': model_psf,
+            'psfShape': psf_shape,
+            'blends': blends,
+        })
+
+    def _model_to_zip(self, model_data: scl.io.ScarletModelData) -> dict[str, Any]:
+        """Convert a ScarletModelData instance to a dictionary of files.
+
+        Parameters
+        ----------
+        model_data : `scl.io.ScarletModelData`
+            ScarletModelData instance.
+
+        Returns
+        -------
+        zip_dict :
+            Dictionary mapping filenames to the desired file type.
+        """
+        json_model = model_data.as_dict()
+
+        data = {
+            f'blend_{blend_id}.json': json.dumps(blend_data)
+            for blend_id, blend_data in json_model['blends'].items()
+        }
+        data.update({
+            'psf_shape.json': json.dumps(json_model['psfShape']),
+            'psf.json': json.dumps(json_model['psf']),
+        })
+        return data
+
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None, expected_size: int = -1) -> Any:
+        # Override of `FormatterV2.read_from_uri`.
+        if self.file_descriptor.parameters is not None and "blend_id" in self.file_descriptor.parameters:
+            blend_id = self.file_descriptor.parameters["blend_id"]
```

Comment:
It seems like the infrastructure here could just as easily support a request for multiple blends in parameters rather than a single blend. Is that at all useful? `blend_ids` could be scalar or list of IDs (and `lsst.utils.iteration.ensure_iterable` can unify them to a single list).